src/javascript/waxeye.ts

```diff
@@ -0,0 +1,734 @@
+/*
+ * Waxeye Parser Generator http://www.waxeye.org
+ * Licensed under the MIT license. See 'LICENSE' for details.
+ */
+
+export class WaxeyeParser {
+  public constructor(
+      public env: ParserConfig,
+      public start: keyof ParserConfig /* keyof env */) {}
+
+  public parse(input: string, start: string = this.start): AST|ParseError {
+    if (!this.env[start]) {
+      throw new Error(`Invalid non-terminal ${start}. Expected one of: ${
+          Object.keys(this.env).join(', ')}`);
+    }
+    return match(this.env, start, input);
+  }
+}
+
+export interface ParserConfig {
+  // name -> NonTerminal
+  [key: string]: NonTerminal;
+}
+
+export interface NonTerminal {
+  mode: NonTerminalMode;
+  exp: Expr;
+}
+
+export const enum NonTerminalMode {
+  NORMAL = 1,
+  PRUNING,
+  VOIDING,
+}
+
+/*
+ * An abstract syntax tree holds the non-terminal's name (`type`),
+ * and a list of child ASTs.
+ */
+export class AST {
+  constructor(public type: string, public children: Array<AST|string>) {}
+
+  /**
+   * The empty AST is an AST that has an empty `type` and no children.
+   */
+  public isEmpty(): this is EmptyAST {
+    return this.type === '' && this.children.length === 0;
+  }
+}
+
+export interface EmptyAST extends AST {
+  type: '';
+  children: never[];
+}
+
+export function EmptyAST(): EmptyAST {
+  return new AST('', []) as EmptyAST;
```

Comment:
Does typescript not infer the type correctly without 'as EmptyAST' ?

Reply:
It does not. The `as` typecast will fail with any other value though (it's not like `static_cast`).